src/tinygp/kernels/quasisep.py

```diff
@@ -635,134 +638,362 @@ def transition_matrix(self, X1: JAXArray, X2: JAXArray) -> JAXArray:
+        the lowest order term.
+    """
+
+    size = quads_coeffs.shape[0] - 1
+    remain = size % 2
+    nPair = size // 2
+    mult_f = quads_coeffs[-1:]  # The coeff of highest order term in the output
+
+    poly = jax.lax.cond(
+        remain == 1,
+        lambda x: jnp.array([1.0, x]),
+        lambda _: jnp.array([0.0, 1.0]),
+        quads_coeffs[-2],
+    )
+    poly = poly[-remain + 1 :]
+
+    for p in jnp.arange(nPair):
+        poly = jnp.convolve(
+            poly,
+            jnp.append(
+                jnp.array([quads_coeffs[p * 2], quads_coeffs[p * 2 + 1]]),
+                jnp.ones((1,)),
+            )[::-1],
+        )
+
+    # the returned is low->high following Kelly+14
+    return poly[::-1] * mult_f
+
+
+def carma_poly2quads(poly_coeffs: JAXArray) -> JAXArray:
+    """Factorize a polynomial into a product of quadratic equations
+
+    Args:
+        poly_coeffs: Coefficients of the input characteristic polynomial. The
+            first entry corresponds to the lowest order term.
+
+    Returns:
+        The 0th and 1st order coefficients of the quadractic equations. The last
+        entry is a multiplier, which corresponds to the coefficient of the highest
+        order term in the full polynomial.
+    """
+
+    quads = jnp.empty(0)
+    mult_f = poly_coeffs[-1]
+    roots = carma_roots(poly_coeffs / mult_f)
+    odd = bool(len(roots) & 0x1)
+
+    rootsComp = roots[roots.imag != 0]
+    rootsReal = roots[roots.imag == 0]
+    nCompPair = len(rootsComp) // 2
+    nRealPair = len(rootsReal) // 2
+
+    for i in range(nCompPair):
+        root1 = rootsComp[i]
+        root2 = rootsComp[i + 1]
+        quads = jnp.append(quads, (root1 * root2).real)
+        quads = jnp.append(quads, -(root1.real + root2.real))
+
+    for i in range(nRealPair):
+        root1 = rootsReal[i]
+        root2 = rootsReal[i + 1]
+        quads = jnp.append(quads, (root1 * root2).real)
+        quads = jnp.append(quads, -(root1.real + root2.real))
+
+    if odd:
+        quads = jnp.append(quads, -rootsReal[-1].real)
+
+    return jnp.append(quads, jnp.array(mult_f))
+
+
+def carma_acvf(arroots: JAXArray, arparam: JAXArray, maparam: JAXArray) -> JAXArray:
+    r"""Compute the coefficients of the autocovariance function (ACVF)
+
+    Args:
+        arroots: The roots of the autoregressive characteristic polynomial.
+        arparam: :math:`\alpha` parameters
+        maparam: :math:`\beta` parameters
+
+    Returns:
+        ACVF coefficients, each entry corresponds to one root.
+    """
+    from jax._src import dtypes
+
+    arparam = jnp.atleast_1d(arparam)
+    maparam = jnp.atleast_1d(maparam)
+
+    complex_dtype = dtypes.to_complex_dtype(arparam.dtype)
+
+    p = arparam.shape[0]
+    q = maparam.shape[0] - 1
+    sigma = maparam[0]
+
+    # normalize beta_0 to 1
+    maparam = maparam / sigma
+
+    # init acf product terms
+    num_left = jnp.zeros(p, dtype=complex_dtype)
```

Comment:
We don't want to specify `jnp.complex128` here because we need to be able to support lower precision versions. The `dtypes.to_complex_dtype(...)` call above should give us the appropriate dtype.

Reply:
Got it, thanks!